### ISSUE TITLE
I'd like to request batch-training support for the MultitaskGPModel. Here is my code attempting to integrate a DataLoader to train a MultitaskGPModel. Unfortunately, I receive the RuntimeError: You must train on the training inputs! [Feature Request]

### ISSUE THREAD

**Issue by ianhill60** [cornellius-gp/gpytorch#1763]
# 🚀 Feature Request

## Motivation

**Is your feature request related to a problem? Please describe.**

## Pitch

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Are you willing to open a pull request?** (We LOVE contributions!!!)

## Additional context

****

**Comment by gpleiss** [comment]
Hi @ianhill60 - batch training for multitask models is currently supported. Without a code example that reproduces your bug, we won't be able to help out.

**Comment by ianhill60** [comment]
Thanks for getting back! Does my error have to do with how I induce the model? 

Here is my code and a data sample for testing: 
https://github.com/ianhill60/gprsampledataset/blob/7de2aa26288e6bd95d1e5772fcab81dc1d165b6a/multitaskBatched.py
https://github.com/ianhill60/gprsampledataset/blob/a52c97f45ddb065dac8e9e32b2ce54c7afec3a02/xdf.csv
https://github.com/ianhill60/gprsampledataset/blob/a52c97f45ddb065dac8e9e32b2ce54c7afec3a02/ydf.csv
The error I get is: 
[tensor(1.5345, dtype=torch.float64, grad_fn=<NegBackward>)]
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-24-888438ae2c38> in <module>()
     93         model.train()
     94         likelihood.train()
---> 95         output = model(x_batch)
     96         loss = -mll(output, y_batch)
     97         loss.backward()

/usr/local/lib/python3.7/dist-packages/gpytorch/models/exact_gp.py in __call__(self, *args, **kwargs)
    254             if settings.debug.on():
    255                 if not all(torch.equal(train_input, input) for train_input, input in zip(train_inputs, inputs)):
--> 256                     raise RuntimeError("You must train on the training inputs!")
    257             res = super().__call__(*inputs, **kwargs)
    258             return res

RuntimeError: You must train on the training inputs!

Thank you for the help!

**Comment by wjmaddox** [comment]
Yes, this is because your dataloader is causing the training inputs to change at each iteration (aka you're trying to use stochastic gradient descent). To be able to not evaluate the entire dataset at each training step, you'd need to use a variational model (aka something like this model https://github.com/cornellius-gp/gpytorch/blob/master/examples/04_Variational_and_Approximate_GPs/SVGP_Multitask_GP_Regression.ipynb).

**Comment by ianhill60** [comment]
Thanks for the helpful response wjmaddox! 
I applied this approach to my problem, but am still running into memory overflow errors even with batching. 
https://github.com/ianhill60/gprsampledataset/blob/69359be7f70c4325f004d2cbcb44ebd4d1fca1dc/VariationalMultitaskApplication.py
What am I missing? I'm having a hard time believing 66k points (12 features and 3 tasks) requires too much memory for batched training. 

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-59a922b67936> in <module>()
    138     # Within each iteration, we will go over each minibatch of data
    139     optimizer.zero_grad()
--> 140     output = model(train_x)
    141     loss = -mll(output, train_y)
    142     epochs_iter.set_postfix(loss=loss.item())

14 frames
/usr/local/lib/python3.7/dist-packages/gpytorch/kernels/kernel.py in _sq_dist(self, x1, x2, postprocess, x1_eq_x2)
     41         x1_ = torch.cat([-2.0 * x1, x1_norm, x1_pad], dim=-1)
     42         x2_ = torch.cat([x2, x2_pad, x2_norm], dim=-1)
---> 43         res = x1_.matmul(x2_.transpose(-2, -1))
     44 
     45         if x1_eq_x2 and not x1.requires_grad and not x2.requires_grad:

RuntimeError: CUDA out of memory. Tried to allocate 63.01 GiB (GPU 0; 11.17 GiB total capacity; 252.02 MiB already allocated; 10.46 GiB free; 276.00 MiB reserved in total by PyTorch)

**Comment by wjmaddox** [comment]
You should be able to sub-sample the data so you don't evaluate on the entire dataset at once. In general, if you have 66k points total, then it looks like you're trying to construct a 66k x 66k kernel matrix which should be at least 17gb (31ish if in double precision).

**Comment by ianhill60** [comment]
Thanks again! I was able to get this to run with a dataloader and batching, however my loss is not decreasing. It fluctuates between around 3.9 - 4.1 regardless of how many epochs I run or changes to my initial learning rate for the adam optimizer (I tried SGD too, was slightly worse). I'm curious if this is because I am updating my optimizer every batch instead of every epoch. Would it be better to only update the optimizer every epoch? 
Are there any mll alternatives to the VariationalELBO that work for batched variational multitask models? 
I'll work on integrating this into an Ax/Botorch workflow to optimize hyperparameters, but I'm worried hyperparameter optimization will be ineffective until I figure out the reason behind the asymptotic learning. 
Here's the code: 
https://github.com/ianhill60/gprsampledataset/blob/b7dd9ac9e84f108b6061f69ca1111d21a5b866dd/VariationalMultitaskApplication.py

Thanks!!

**Comment by gpleiss** [comment]
Try z-scoring your data (make each `y` dimension mean zero and standard deviation 1, same for each `x` dimension). We have found this generally improves optimization and makes everything more stable.

**Comment by gpleiss** [comment]
Oops I'm sorry - I just saw that you are doing that in your code! I missed that on my first pass.

**Comment by gpleiss** [comment]
A few ideas:
1. Try using more than 20 inducing points. I would use at least 300 - ideally at least 1000
2. Rather than randomly initializing the inducing locations, fit them to your data using kmeans clustering.
3. Adam should be much better at optimization than SGD.

**Comment by ianhill60** [comment]
Thanks gpleiss! I implemented kmeans to learn inducing locations and used 1000 points. Model results improved but asymptote quickly (the loss starts at about 7, gets to a loss of ~3.6/3.7 in 100 epochs, and ends at a loss of 3.53 at 1000 epochs). Here is my updated code: https://github.com/ianhill60/gprsampledataset/blob/5982ea5d991ee6fb733627211dd60e447a6750e5/VariationalMultitaskApplication.py
I'm not really sure how to improve from here, so I will continue implementing the bayesian optimization for hyperparameters. The dataset I'm working with might just be difficult to model. 
Thanks!

**Comment by gpleiss** [comment]
My only other suggestion would be to use a slightly larger learning rate with Adam. Otherwise, everything looks reasonable in your code (at least at a quick glance).

The fact that the loss is dropping from 7 to 3.5 suggests that it is no longer an optimization issue. My best guess then for why it gets stuck there is - as you suggest - your dataset is difficult. Not sure we can offer much more help here.

**Comment by ianhill60** [comment]
Thank you gpleiss!
I noticed that increasing the number of latent functions learned to correlate tasks increased my prediction accuracy and need for more data. Is there a 'right' number of latents to use? Am I just overfitting? Is there a good way to evaluate overfitting in Gpytorch?
